### PR TITLE
fix ConvertDecimalType unexpected behavior

### LINKED ISSUE #33906: Integer from VALUES cannot be parsed as Decimal256
`SELECT * FROM values('x Decimal256(0)', (1))`

```
Code: 53. DB::Exception: Type mismatch in IN or VALUES section. Expected: Decimal(76, 0). Got: UInt64. (TYPE_MISMATCH)
```

### PR DESCRIPTION
**Pull request by awakeljw**
### Changelog category (leave one):

- Bug Fix (user-visible misbehaviour in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
close #33906

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: tests/queries/0_stateless/01658_values_ubsan.sql [hunk 1]
```diff
@@ -1 +1,5 @@
 SELECT * FROM VALUES('x UInt8, y UInt16', 1 + 2, 'Hello'); -- { serverError 36 }
+SELECT * FROM VALUES('x Decimal32(0)', (1));
```
Comment (kssenii) [inline tests/queries/0_stateless/01658_values_ubsan.sql:2]: Let's put it in a separate test file

Comment (awakeljw) [reply tests/queries/0_stateless/01658_values_ubsan.sql]: OK, I'll fix it.

File: src/Interpreters/convertFieldToType.cpp [hunk 2]
```diff
@@ -190,9 +190,14 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
...
-        if (const auto * ptype = typeid_cast<const DataTypeDecimal<Decimal64> *>(&type)) return convertDecimalType(src, *ptype);
-        if (const auto * ptype = typeid_cast<const DataTypeDecimal<Decimal128> *>(&type)) return convertDecimalType(src, *ptype);
+#define M(TYPE) \
+        if (const auto * ptype = typeid_cast<const DataTypeDecimal<TYPE> *>(&type)) return convertDecimalType(src, *ptype);
+
+        M(Decimal32)
+        M(Decimal64)
+        M(Decimal128)
+        M(Decimal256)
+#undef M
```
Comment (kssenii) [inline src/Interpreters/convertFieldToType.cpp:200]: Let's leave a previous version before force push when just 1 line was added with the `if` for Decimal256.

Comment (awakeljw) [reply src/Interpreters/convertFieldToType.cpp]: OK, I have fixed it.

**Review by kssenii — APPROVED** [review]

**Pull request merged**